specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+
+#### TYPE_NEGATIVE_FLOAT
+
+```
+| 4 | MATISSA_LENGTH | VARINT
+```
+
+#### TYPE_STRING
+
+```
+| 5 | VARINT_LENGTH | STRING
+```
+
+Note: This is essentially the same as Binary, but with a different token flag.
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged. See [Strings].
+
+#### TYPE_BINARY
+
+```
+| 6 | VARINT_LENGTH | BINARY
+```
+
+#### TYPE_MAP
+
+```
+| 7 | VARINT_NUM_PAIRS | PAIRS
+```
+
+`PAIRS` contains alternating keys then values concatenated. i.e. ` KEY1 | VALUE1 | KEY2 | VALUE2 ...`
+
+This codec does not have any form of canonical map sorting as that would make it ill-fitted to the IPLD Data Model.
+
+As in the IPLD Data Model map keys must be of type String, however as described in the String section the only distinction between Strings and Bytes are identifier hints.
+
+Note: Simple-DAG went with `KEYS_VARINT_LENGTH | VALUES_VARINT_LENGTH | KEYS | VALUES |`. Both seem doable,
+this approach seemed to make writing encoder/decoders really simple. However, adding in more length prefixes
+makes creating faster "zero copy" decoders very nice. It seems to mostly be a tradeoff for which side has to have bigger buffers, the encoder or the decoder.
+
+Note: We could assert that keys are just `VARINT_LENGTH | STRING` and remove the String token since it's always a string.
+It's some added complexity and really stops us from putting anything other than Strings in map keys, but that may not be too bad.
+
+#### TYPE_LIST
+
+```
+| 8 | VARINT_NUM_ELEMENTS | VALUES |
+```
+
+`VALUES` contains every value concatenated.
+
+Note: Simple-DAG went with `VARINT_LENGTH` (the size of the VALUES binary section) instead of `VARINT_NUM_ELEMENTS`
+and in the `VALUES` section had every value proceeded by the length of the value.
+
+Both seem doable, this approach seemed to make writing a decoder really simple. However, adding in more length prefixes makes creating faster "zero copy" decoders very nice. It seems to mostly be a tradeoff for which side has to have bigger buffers, the encoder or the decoder.
+
+#### TYPE_NULL
+
+```
+| 9 |
+```
+
+#### TYPE_TRUE
+
+```
+| 10 |
+```
+
+#### TYPE_FALSE
+
+```
+| 11 |
+```
+
+### Strings
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged.
+
+## Implementations
+
+### Go
+
+Here and adheres to the specification. However, in a practical sense because it implements the go-ipld-prime interface and its limits on integers (and floats once specified) are currently bounded by those interfaces. Similarly any varints are capped around 64 bits.
+
+### Rust/WASM
+
+Adheres to the specification. However, in a practical sense its limits on integers and floats are currently bounded to be of fixed maximum sizes. Similarly, all varints are capped around 64 bits.
```

Comment:
To enable complex data passing over the Wasm boundary, MsgPack (or another serialization standard) can be used along with a standardized ABI. This also comes with the benefit of being language agnostic.

At the [Polywrap](https://polywrap.io/) project we've had success with this approach, and now have support for building composable wasm modules in AssemblyScript and Rust (TinyGo OTW). Might be useful to check it out for inspiration. (core standard being defined [here](https://github.com/polywrap/specification/blob/main/standard/01.0%20General%20Standard/WRAP%20Standard.md)

Reply:
CBOR could be a good option, and we've been using it in the FVM with some success (though mostly tuple encoded structures)